akka-http/src/main/scala/akka/http/scaladsl/server/ContentNegotation.scala

```diff
@@ -0,0 +1,265 @@
+    } else None
+}
+
+object ContentNegotiator {
+  sealed trait Alternative {
+    def mediaType: MediaType
+    def format: String
+  }
+  object Alternative {
+    implicit def apply(contentType: model.ContentType): ContentType = ContentType(contentType)
+    implicit def apply(mediaType: model.MediaType): Alternative =
+      mediaType match {
+        case x: model.MediaType.Binary           ⇒ ContentType(x)
+        case x: model.MediaType.WithFixedCharset ⇒ ContentType(x)
+        case x: model.MediaType.WithOpenCharset  ⇒ MediaType(x)
+      }
+
+    case class ContentType(contentType: model.ContentType) extends Alternative {
+      def mediaType = contentType.mediaType
+      def format = contentType.toString
+    }
+    case class MediaType(mediaType: model.MediaType.WithOpenCharset) extends Alternative {
+      def format = mediaType.toString
+    }
+  }
+
+  def apply(requestHeaders: Seq[HttpHeader]) = new ContentNegotiator(requestHeaders)
+}
+
+final class EncodingNegotiator(requestHeaders: Seq[HttpHeader]) {
+
+  /**
+   * The encoding-ranges accepted by the client according to given request headers, sorted by
+   * 1. increasing generality (i.e. most specific first)
+   * 2. decreasing q-value (only for ranges targeting a single HttpEncoding)
+   * 3. order of appearance in the `Accept-Encoding` header(s)
+   */
+  val acceptedEncodingRanges: List[HttpEncodingRange] =
+    (for {
+      `Accept-Encoding`(encodingRanges) ← requestHeaders
+      range ← encodingRanges
+    } yield range).sortBy { // `sortBy` is stable, i.e. upholds the original order on identical keys
+      case _: HttpEncodingRange.`*` ⇒ 1f // most general, needs to come last
+      case x                        ⇒ -x.qValue // all others come first
+    }.toList
+
+  /**
+   * Returns the q-value that the client (implicitly or explicitly) attaches to the given encoding.
+   * See http://tools.ietf.org/html/rfc7231#section-5.3.1 for details.
+   */
+  def qValueFor(encoding: HttpEncoding): Float =
+    acceptedEncodingRanges match {
+      case Nil ⇒ 1.0f
+      case x   ⇒ x collectFirst { case r if r matches encoding ⇒ r.qValue } getOrElse 0f
+    }
+
+  /**
+   * Determines whether the given encoding is accepted by the client.
+   */
+  def isAccepted(encoding: HttpEncoding): Boolean = qValueFor(encoding) > 0f
+
+  /**
+   * Determines whether the request has an `Accept-Encoding` clause matching the given encoding.
+   */
+  def hasMatchingFor(encoding: HttpEncoding): Boolean =
+    acceptedEncodingRanges.exists(_ matches encoding)
+
+  /**
+   * Picks the best of the given encoding alternatives given the preferences
+   * the client indicated in the request's `Accept-Encoding` headers.
+   * See http://tools.ietf.org/html/rfc7231#section-5.3.4 for details on the negotiation logic.
+   *
+   * If there are several best encoding alternatives that the client has equal preference for
+   * the order of the arguments is used as a tie breaker (first one wins).
```

Comment:
> So the stable sort in acceptedXXX doesn't really matter, right?

No any more it doesn't, no.
In the first version I followed the spec by the book, which says that the order of the ranges in the Accept* headers should break ties. However, if we implement it this way then the application developer has no way to establish her own precedence, which is something I'd value higher than strictly following the spec in this matter.

So, if the client says he'll accept both `deflate` and `gzip` with equal preference then we can decide what to pick via our alternative ordering. If the client really wants `gzip` rather than `deflate` he needs to signal that preference with q-values.